### ISSUE TITLE
[Feature] [e2e-infra] Support duplicate data varification

### ISSUE THREAD

**Issue by wu-sheng** [apache/skywalking#12253]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In the 0.6 BanyanDB release, we missed an important feature reported by #12219.

The list data verification should include whether there is duplicate data.

@fgksgf @kezhenxu94  Could you suggest what could be done and how to verify?

### Are you willing to submit a pull request to implement this on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by zzzk1** [comment]
Hi @yswdqz, are you still working on this? If not, can I pick up this issue, or I can go ahead and push my idea forward.

**Comment by yswdqz** [comment]
@zzzk1  Of course, you're welcome to participate if you're interested, and if you have any questions, feel free to ask me.

**Comment by yswdqz** [comment]
@zzzk1 Hello, is there any progress, or is there something blocking?

**Comment by yswdqz** [comment]
If @zzzk1 no longer spends time on this, I will work on it starting from October 1.

**Comment by wu-sheng** [comment]
Changing the owner back to you. I don't see update for weeks.

**Comment by amethystani** [comment]
Hey can you assign me this issue. I see there is no progress

**Comment by nikumbgo12345** [comment]
Not sure if this is still required, but I would like to give this a shot.

**Comment by kezhenxu94** [comment]
> Not sure if this is still required, but I would like to give this a shot.

Please go ahead.

**Comment by aryanttripathi** [comment]
Hi @wu-sheng @kezhenxu94 ,
Can I work on this issue?

I went through the discussion on the related thread — [#12219](https://github.com/apache/skywalking/issues/12219). However, as I’m new to SkyWalking, it’s a bit challenging for me to fully grasp the context.

Could you please help me understand what exactly needs to be fixed and how I can reproduce the issue (if applicable)?
It would also be great if you could suggest some relevant entry points in the codebase to start exploring.

I’m really interested in contributing to SkyWalking, but this issue seems quite broad for a newcomer, so any guidance would be appreciated.

Thank you!

**Comment by kezhenxu94** [comment]
@Aryant-Tripathi I think @Navadeep0007 has already created a PR there

**Comment by wu-sheng** [comment]
The `noDuplicates` pipe function has been implemented in apache/skywalking-infra-e2e#145.

It can be used with `contains`, `containsOnce`, or `range` via standard Go template pipe syntax:

```yaml
{{- contains (.metrics | noDuplicates) }}
- name: {{ notEmpty .name }}
  value: {{ notEmpty .value }}
{{- end }}
```

When the actual data contains duplicate entries (full deep equality), verification will fail with an error indicating the duplicated item.

This PR also builds on top of two earlier fixes:
- apache/skywalking-infra-e2e#143 — fixed `contains` to enforce distinct matching (each actual item can only satisfy one expected entry), and added `containsOnce` with backtracking for optimal assignment.

**Comment by wu-sheng** [comment]
Once https://github.com/apache/skywalking-infra-e2e/pull/145 is merged and infra-e2e 1.4.0 is released, we plan to enhance BanyanDB's e2e tests by adding `| noDuplicates` pipe to list-type verifications in expected YAML files.

The primary targets are the cases directly related to the original duplicate bug reported in #12219:

- `service-instance-provider.yml` / `service-instance-consumer.yml` — service instance list queries
- `metrics-top-endpoint.yml` / `metrics-top-service.yml` — topN query results

For example, the expected file would change from:
```yaml
{{- contains . }}
- id: {{ b64enc "e2e-service-provider" }}.1_{{ b64enc "provider1" }}
  name: provider1
  ...
{{- end}}
```
to:
```yaml
{{- contains (. | noDuplicates) }}
- id: {{ b64enc "e2e-service-provider" }}.1_{{ b64enc "provider1" }}
  name: provider1
  ...
{{- end}}
```

We will also apply it to other list verifications including traces list, dependency nodes/calls, and event lists across all e2e test suites (storage, lifecycle, profiling, log, event, zipkin).
